cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -38,24 +37,110 @@ trait Traverser extends Iterator[Node] { self =>
  * Bounds an iterator to go no more than a specified maximum number of steps
  */
 trait BoundedIterator[T] extends Iterator[T] {
-  val maxSteps: Long
+  def maxSteps: Option[Long]
+
   private var numStepsTaken = 0L
 
-  abstract override def next = {
+  abstract override def next() = {
     numStepsTaken += 1
-    super.next
+    super.next()
   }
 
-  abstract override def hasNext = ( (numStepsTaken < maxSteps) && super.hasNext)
+  abstract override def hasNext = maxSteps match {
+    case Some(max) if max > numStepsTaken => super.hasNext
+    case _ => Iterator.empty.hasNext
+  }
 }
 
 /**
- * Randomly traverse the graph, going from one node to a random neighbor in direction {@code dir}.
+ * General schema for most of Traversers. `QueueBasedTraverser` keeps nodes to
+ * visit next in a queue. It iteratively visits a node from the top of the
+ * queue and optionally adds new nodes to the queue.
+ */
+trait QueueBasedTraverser extends Traverser {
+  /**
+   * Graph to be traversed.
+   */
+  def graph: Graph
+
+  /**
+   * Nodes to start the traversal from.
+   */
+  def homeNodeIds: Seq[Int]
+
+  /**
+   * The priority of nodes adding to the queue. `BFS` adds nodes at the
+   * end and `DFS` adds nodes at the beginning.
+   */
+  object GraphTraverserNodePriority extends Enumeration {
+    type GraphTraverserNodePriority = Value
+
+    val FIFO = Value
+    val LIFO = Value
+  }
+
+  import GraphTraverserNodePriority._
+
+  /**
+   * The priority of nodes adding to the queue.
+   */
+  def nodePriority: GraphTraverserNodePriority
+
+  /**
+   * Action performed when visiting node `node` (before iteratur retuns `node`).
+   * Should be implemented in subclass.
+   */
+  def visitNode(node: Node): Unit = {}
```

Comment:
The short notation will be deprecated in near future:
https://issues.scala-lang.org/browse/SI-7605
I suggest leaving it everywhere.